File: sdks/kotlin/src/main/kotlin/com/getbindu/sdk/CoreLauncher.kt
/**
 * Core Launcher — spawns the Bindu Python core as a child process.
 *
 * Same pattern as the TypeScript SDK: detect bindu CLI, spawn it,
 * wait for the gRPC port to be ready.
 */

package com.getbindu.sdk

import java.io.BufferedReader
import java.io.InputStreamReader
import java.net.Socket

object CoreLauncher {

    /**
     * Launch the Bindu Python core.
     *
     * Tries in order:
     *   1. `bindu serve --grpc`
     *   2. `uv run bindu serve --grpc`
     *   3. `python3 -m bindu.cli serve --grpc`
     *
     * @param grpcPort gRPC server port (default: 3774).
     * @param httpPort HTTP server port (default: 3773).
     * @param timeoutMs Maximum time to wait for core to start.
     * @return The child process, or null if core is already running.
     */
    fun launch(
        grpcPort: Int = 3774,
        httpPort: Int = 3773,
        timeoutMs: Long = 30000
    ): Process? {
        // Check if core is already running
        if (isPortOpen(grpcPort)) {
            println("  Bindu core already running on :$grpcPort")
            return null
        }

        val (command, args) = findCommand(grpcPort, httpPort)
        println("  Starting Bindu core: $command ${args.joinToString(" ")}")

        val processBuilder = ProcessBuilder(listOf(command) + args)
            .redirectErrorStream(true)

        val process = processBuilder.start()

        // Pipe output with prefix
        Thread {
            BufferedReader(InputStreamReader(process.inputStream)).use { reader ->
                reader.lines().forEach { line ->
                    println("[bindu-core] $line")
                }
            }
        }.apply { isDaemon = true }.start()

        // Wait for gRPC port to be ready
        println("  Waiting for Bindu core gRPC on port $grpcPort...")
        waitForPort(grpcPort, timeoutMs)
        println("  Bindu core is ready.")

        return process
    }

    private fun findCommand(grpcPort: Int, httpPort: Int): Pair<String, List<String>> {
        val args = listOf("serve", "--grpc", "--grpc-port", grpcPort.toString(), "--port", httpPort.toString())

        // Try bindu CLI
        if (commandExists("bindu")) {
            return "bindu" to args
        }

        // Try uv
        if (commandExists("uv")) {
            return "uv" to listOf("run", "bindu") + args
        }

        // Fallback to python3
        return "python3" to listOf("-m", "bindu.cli") + args
    }

    private fun commandExists(command: String): Boolean {
        return try {
            val process = ProcessBuilder("which", command)
                .redirectErrorStream(true)
                .start()
            process.waitFor() == 0
        } catch (_: Exception) {
            false
        }
    }

    private fun isPortOpen(port: Int, host: String = "localhost"): Boolean {
        return try {
            Socket(host, port).use { true }
        } catch (_: Exception) {
            false
        }
    }

    private fun waitForPort(port: Int, timeoutMs: Long) {
        val start = System.currentTimeMillis()
        while (System.currentTimeMillis() - start < timeoutMs) {
            if (isPortOpen(port)) return
            Thread.sleep(500)
        }
        throw RuntimeException("Bindu core did not start within ${timeoutMs / 1000}s on port $port")
    }
}

File: sdks/typescript/src/core-launcher.ts
/**
 * Core Launcher — spawns the Bindu Python core as a child process.
 *
 * When a TypeScript developer calls bindufy(), the SDK needs the Python core
 * running to handle DID, auth, x402, A2A protocol, scheduler, and storage.
 * This module detects if Bindu is installed, spawns it as a child process,
 * and waits for the gRPC server to be ready.
 *
 * The developer sees one command, one terminal — the child process is hidden.
 */

import { spawn, ChildProcess } from "child_process";
import * as net from "net";

/** Check if a port is open (gRPC server is ready). */
function waitForPort(
  port: number,
  host: string = "localhost",
  timeoutMs: number = 30000
): Promise<void> {
  return new Promise((resolve, reject) => {
    const startTime = Date.now();

    const tryConnect = () => {
      const socket = new net.Socket();
      socket.setTimeout(1000);

      socket.on("connect", () => {
        socket.destroy();
        resolve();
      });

      socket.on("error", () => {
        socket.destroy();
        if (Date.now() - startTime > timeoutMs) {
          reject(
            new Error(
              `Bindu core did not start within ${timeoutMs / 1000}s on port ${port}`
            )
          );
        } else {
          setTimeout(tryConnect, 500);
        }
      });

      socket.on("timeout", () => {
        socket.destroy();
        setTimeout(tryConnect, 500);
      });

      socket.connect(port, host);
    };

    tryConnect();
  });
}

/** Find the bindu executable (checks pip-installed bindu CLI). */
function findBinduExecutable(): string | null {
  const { execSync } = require("child_process");
  try {
    const result = execSync("which bindu 2>/dev/null || where bindu 2>nul", {
      encoding: "utf-8",
    }).trim();
    return result || null;
  } catch {
    return null;
  }
}

/** Check if uv is available for running Python. */
function hasUv(): boolean {
  const { execSync } = require("child_process");
  try {
    execSync("which uv 2>/dev/null || where uv 2>nul", { encoding: "utf-8" });
    return true;
  } catch {
    return false;
  }
}

/**
 * Launch the Bindu Python core as a child process.
 *
 * Tries in order:
 *   1. `bindu serve --grpc` (pip-installed CLI)
 *   2. `uv run bindu serve --grpc` (uv-managed)
 *   3. `python -m bindu.cli serve --grpc` (fallback)
 *
 * @param grpcPort - Port for the gRPC server (default: 3774).
 * @param httpPort - Port for the HTTP A2A server (default: 3773).
 * @returns The child process and the actual ports.
 */
export async function launchCore(
  grpcPort: number = 3774,
  httpPort: number = 3773
): Promise<{ process: ChildProcess; grpcPort: number; httpPort: number }> {
  let command: string;
  let args: string[];

  const binduPath = findBinduExecutable();
  if (binduPath) {
    command = binduPath;
    args = ["serve", "--grpc", "--grpc-port", String(grpcPort)];
  } else if (hasUv()) {
    command = "uv";
    args = ["run", "bindu", "serve", "--grpc", "--grpc-port", String(grpcPort)];
  } else {
    command = "python3";
    args = ["-m", "bindu.cli", "serve", "--grpc", "--grpc-port", String(grpcPort)];
  }

  console.log(`Starting Bindu core: ${command} ${args.join(" ")}`);

  const child = spawn(command, args, {
    stdio: ["ignore", "pipe", "pipe"],
    detached: false,
  });

  // Pipe core output to console with prefix
  child.stdout?.on("data", (data: Buffer) => {
    const lines = data.toString().trim().split("\n");
    lines.forEach((line: string) => console.log(`[bindu-core] ${line}`));
  });

  child.stderr?.on("data", (data: Buffer) => {
    const lines = data.toString().trim().split("\n");
    lines.forEach((line: string) => console.error(`[bindu-core] ${line}`));
  });

  child.on("error", (err: Error) => {
    console.error(
      `Failed to start Bindu core: ${err.message}\n` +
        "Make sure Bindu is installed: pip install bindu[grpc]"
    );
    process.exit(1);
  });

  child.on("exit", (code: number | null) => {
    if (code !== null && code !== 0) {
      console.error(`Bindu core exited with code ${code}`);
    }
  });

  // Kill child process when parent exits
  const cleanup = () => {
    if (!child.killed) {
      child.kill("SIGTERM");
    }
  };
  process.on("exit", cleanup);
  process.on("SIGINT", () => {
    cleanup();
    process.exit(0);
  });
  process.on("SIGTERM", () => {
    cleanup();
    process.exit(0);
  });

  // Wait for gRPC server to be ready
  console.log(`Waiting for Bindu core gRPC on port ${grpcPort}...`);
  await waitForPort(grpcPort);
  console.log("Bindu core is ready.");

  return { process: child, grpcPort, httpPort };
}
